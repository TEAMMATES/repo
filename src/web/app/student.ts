/**
 * A student.
 */
export interface Student {
  name: string;
  lastName: string;
}

/**
 * A student's response status for a feedback session.
 */
export interface StudentResponseStatus {
  email: string;
  name: string;
  sectionName: string;
  teamName: string;
  responseStatus: boolean;
}

/**
 * All the students' response status for a feedback session.
 */
export interface StudentsResponseStatus {
  studentsResponseStatus: StudentResponseStatus[];
}
