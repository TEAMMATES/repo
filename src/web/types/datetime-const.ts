export enum HoursConst {
    ZERO_HOURS = 0,
    TWELVE_HOURS = 12,
    ONE_DAY_HOURS = 24,
    THREE_DAYS_HOURS = 72,
    ONE_WEEK_HOURS = 168,
}

export enum SecondsConst {
    ONE_MINUTE_SECONDS = 60,
}

export enum MinutesConst {
    ONE_HOURS_MINUTES = 60,
}

export enum MsConst {
    ONE_SECOND_MILLISECONDS = 1000,
    TEN_SECOND_MILLISECONDS = 10000,
    ONE_MINUTE_MILLISECONDS = 60000,
    TEN_MINUTE_MILLISECONDS = 600000,
    FIFTEEN_MINUTE_MILLISECONDS = 900000,
    ONE_HOUR_MILLISECONDS = 3600000,
    ONE_WEEK_MILLISECONDS = 604800000,
}

export const getDefaultDateFormat = (): DateFormat => {
    return { year: 0, month: 0, day: 0 };
};

export const getLatestTimeFormat = (): TimeFormat => {
    return { hour: 23, minute: 59 };
};

export const getDefaultTimeFormat = (): TimeFormat => {
    return { hour: 0, minute: 0 };
};

/**
 * The date format.
 */
export interface DateFormat {
    year: number;
    month: number;
    day: number;
}

/**
 * The output format of the time picker.
 */
export interface TimeFormat {
    hour: number;
    minute: number;
}
