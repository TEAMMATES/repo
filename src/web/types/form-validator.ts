import { ApiConst, ApiStringConst } from './api-const';

/**
 * Represents the root FormValidator object of all form fields.
 */
export enum FormValidator {
  /**
   * Max length for the 'Course ID' field.
   */
  COURSE_ID_MAX_LENGTH = ApiConst.COURSE_ID_MAX_LENGTH,

  /**
   * Max length for the 'Course Name' field.
   */
  COURSE_NAME_MAX_LENGTH = ApiConst.COURSE_NAME_MAX_LENGTH,

  /**
   * Max length for the 'Student Name` field.
   */
  STUDENT_NAME_MAX_LENGTH = ApiConst.STUDENT_NAME_MAX_LENGTH,

  /**
   * Max length for the 'Section Name` field.
   */
  SECTION_NAME_MAX_LENGTH = ApiConst.SECTION_NAME_MAX_LENGTH,

  /**
   * Max length for the 'Team Name` field.
   */
  TEAM_NAME_MAX_LENGTH = ApiConst.TEAM_NAME_MAX_LENGTH,

  /**
   * Max length for the 'E-mail Address` field.
   */
  EMAIL_MAX_LENGTH = ApiConst.EMAIL_MAX_LENGTH,

  /**
   * Regex used to verify URLs as Angular does not have a built-in URL validator.
   */
  URL_REGEX = ApiStringConst.URL_REGEX,

  /**
   * Regex used to verify emails in the back-end.
   */
  EMAIL_REGEX = ApiStringConst.EMAIL_REGEX,
}
