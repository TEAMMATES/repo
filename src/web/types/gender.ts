/**
 * Represents the gender of a student.
 */
export enum Gender {
  MALE,
  FEMALE,
  OTHER,
}
export namespace Gender {

  export const values: () => string[] = (): string[] =>
      Object.keys(Gender).filter((key: string) => isNaN(key as any) && key !== 'values');

  export const enums: () => Gender[] = (): Gender[] => {
    const genders: Gender[] = [];
    genders.push(Gender.MALE);
    genders.push(Gender.FEMALE);
    genders.push(Gender.OTHER);
    return genders;
  };
}
