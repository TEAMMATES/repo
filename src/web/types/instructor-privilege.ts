import { InstructorPrivilege } from './api-output';

/**
 * The default instructor privilege.
 */
export const DEFAULT_INSTRUCTOR_PRIVILEGE: InstructorPrivilege = {
  canEditCourse: false,
  canDeleteCourse: false,
  canRestoreCourse: false,
  canModifySession: false,
  canModifyStudent: false,
  canSubmitSessionInSections: false,
  canModifyInstructor: false,
  canViewStudentInSections: false,
  canModifySessionCommentsInSections: false,
  canViewSessionInSections: false,
};
