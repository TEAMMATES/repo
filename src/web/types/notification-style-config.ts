import { NotificationStyle } from './api-request';

enum NotificationFontColor {
    WHITE = 'text-white',
    DARK = 'text-dark',
}

interface NotificationStyleConfig {
    font: NotificationFontColor,
    description: string,
}

/**
 * Maps between notification style and actual class/description to be used.
 *
 * Ref: https://getbootstrap.com/docs/4.5/utilities/colors/#background-color
 */
export const notificationStyleConfigMap: Record<NotificationStyle, NotificationStyleConfig> = {
    PRIMARY: { font: NotificationFontColor.WHITE, description: 'Primary (blue)' },
    SECONDARY: { font: NotificationFontColor.WHITE, description: 'Secondary (grey)' },
    SUCCESS: { font: NotificationFontColor.WHITE, description: 'Success (green)' },
    DANGER: { font: NotificationFontColor.WHITE, description: 'Danger (red)' },
    WARNING: { font: NotificationFontColor.DARK, description: 'Warning (yellow)' },
    INFO: { font: NotificationFontColor.WHITE, description: 'Info (cyan)' },
    LIGHT: { font: NotificationFontColor.DARK, description: 'Light' },
    DARK: { font: NotificationFontColor.WHITE, description: 'Dark' },
};
