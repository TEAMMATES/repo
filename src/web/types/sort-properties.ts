/**
 * Elements that tables are sorted by.
 */
export enum SortBy {
    /**
     * Nothing.
     */
    NONE,

    /**
     * Section Name.
     */
    SECTION_NAME,

    /**
     * Team name.
     */
    TEAM_NAME,

    /**
     * Student Name.
     */
    STUDENT_NAME,

    /**
     * The email of the student.
     */
    EMAIL,

    /**
     * The gender of the student.
     */
    STUDENT_GENDER,

    /**
     * Institution.
     */
    INSTITUTION,

    /**
     * Nationality.
     */
    NATIONALITY,

    /**
     * Join status
     */
     JOIN_STATUS,

    /**
     * Course ID.
     */
    COURSE_ID,

    /**
     * Course Name.
     */
    COURSE_NAME,

    /**
     * The creation time of the course.
     */
    COURSE_CREATION_DATE,

    /**
     * Completion status of feedback session.
     */
    SESSION_COMPLETION_STATUS,

    /**
     * Feedback session name.
     */
    SESSION_NAME,

    /**
     * Start time of the feedback session.
     */
    SESSION_START_DATE,

    /**
     * End time of the feedback session.
     */
    SESSION_END_DATE,

    /**
     * The creation time of the feedback session.
     */
    SESSION_CREATION_DATE,

    /**
     * The time when the feedback session is moved to recycle bin.
     */
    SESSION_DELETION_DATE,

    /**
     * Feedback question type.
     */
    QUESTION_TYPE,

    /**
     * Feedback question text (brief).
     */
    QUESTION_TEXT,

    /**
     * Team of the giver of the feedback response.
     */
    GIVER_TEAM,

    /**
     * Name of the giver of the feedback response.
     */
    GIVER_NAME,

    /**
     * Team of the recipient of the feedback response.
     */
    RECIPIENT_TEAM,

    /**
     * Name of the recipient of the feedback response.
     */
    RECIPIENT_NAME,
}

/**
 * Sort order.
 */
export enum SortOrder {
    /**
     * Descending sort order.
     */
    DESC,

    /**
     * Ascending sort order
     */
    ASC,
}
