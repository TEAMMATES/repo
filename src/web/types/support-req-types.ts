export interface SupportRequest extends SupportRequestRequest {
    trackingId: string,
    status: SupportReqStatus
}

export interface SupportRequestRequest {
    email: string,
    name: string,
    type: SupportReqEnquiryType,
    title: string,
    initial_msg: string,
    createdAt: number,
    updatedAt: number,
}

/**
 * The ordering in the supportRequestList is based on the order of this list
 */
export enum SupportReqStatus {
    NEW,
    AWAITING_ADMIN,
    AWAITING_USER,
    RESOLVED
}

/**
 * The ordering in the supportRequestList is based on the order of this list
 */
export enum SupportReqEnquiryType {
    NEW_ACCOUNT,
    GENERAL_HELP
}