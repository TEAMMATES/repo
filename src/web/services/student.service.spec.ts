import { TestBed } from '@angular/core/testing';

import { HttpClientTestingModule } from '@angular/common/http/testing';
import { Observable, of } from 'rxjs';
import { ResourceEndpoints } from '../types/api-endpoints';
import { JoinState, Student } from '../types/api-output';
import { StudentUpdateRequest } from '../types/api-request';
import { HttpRequestService } from './http-request.service';
import { StudentService } from './student.service';
import Spy = jasmine.Spy;

const defaultStudentUpdateRequest: StudentUpdateRequest = {
  name: 'John Doe',
  email: 'johndoe@gmail.com',
  team: '',
  section: '',
  comments: '',
  isSessionSummarySendEmail: true,
};

describe('StudentService', () => {
  let spyHttpRequestService: any;
  let service: StudentService;

  beforeEach(() => {
    TestBed.configureTestingModule({
      imports: [
        HttpClientTestingModule,
      ],
      providers: [HttpRequestService],
    });
    service = TestBed.get(StudentService);
    spyHttpRequestService = TestBed.get(HttpRequestService);
  });

  it('should be created', () => {
    expect(service).toBeTruthy();
  });

  it('should execute PUT when updating students in a course', () => {
    const paramMap: Record<string, string> = {
      courseid: 'CS3281',
      studentemail: 'johndoe@gmail.com',
    };
    spyOn(spyHttpRequestService, 'put').and.stub();

    service.updateStudent({
      courseId: paramMap.courseid,
      studentEmail: paramMap.studentemail,
      requestBody: defaultStudentUpdateRequest,
    });

    expect(spyHttpRequestService.put)
        .toHaveBeenCalledWith(ResourceEndpoints.STUDENT, paramMap, defaultStudentUpdateRequest);
  });

  it('should execute DELETE when deleting all students in a course', () => {
    const paramMap: Record<string, string> = {
      courseid: 'CS3281',
    };
    spyOn(spyHttpRequestService, 'delete').and.stub();

    service.deleteAllStudentsFromCourse({
      courseId: paramMap.courseid,
    });

    expect(spyHttpRequestService.delete)
        .toHaveBeenCalledWith(ResourceEndpoints.STUDENTS, paramMap);
  });

  it('should execute POST when regenerating key of a student in a course', () => {
    const paramMap: Record<string, string> = {
      courseid: 'CS3281',
      studentemail: 'johndoe@gmail.com',
    };
    spyOn(spyHttpRequestService, 'post').and.stub();

    service.regenerateStudentCourseLinks(paramMap.courseid, paramMap.studentemail);

    expect(spyHttpRequestService.post)
        .toHaveBeenCalledWith(ResourceEndpoints.STUDENT_COURSE_LINKS_REGENERATION, paramMap);
  });

  it('should execute GET on course & students endpoints when generating student csv list', () => {
    const courseId: string = 'CS3281';
    const httpSpy: Spy = spyOn(spyHttpRequestService, 'get').and.returnValue(of(''));
    const studentList: Observable<string> = service.loadStudentListAsCsv({ courseId });
    expect(httpSpy).toBeCalledWith(ResourceEndpoints.COURSE, {
      courseid: courseId,
      entitytype: 'instructor',
    });
    studentList.subscribe(() => expect(httpSpy).toBeCalledWith(ResourceEndpoints.STUDENTS, {
      courseid: courseId,
    }));
  });

  it('should show course student list with section', () => {
    const courseId: string = 'listWithSection';
    const courseName: string = 'List With Section';
    const students: Student[] =
      [
        {
          courseId,
          email: 'student1OfTypicalCourse@typicalCourse.com',
          name: 'student1OfTypicalCourse',
          joinState: JoinState.JOINED,
          teamName: 'Team 1',
          sectionName: 'Section A',
        },
        {
          courseId,
          email: 'student2OfTypicalCourse@typicalCourse.com',
          name: 'student2OfTypicalCourse',
          joinState: JoinState.JOINED,
          teamName: 'Team 2',
          sectionName: 'Section A',
        },
      ];
    expect(service.processStudentsToCsv(courseId, courseName, students)).toMatchSnapshot();
  });

  it('should show course student list without section', () => {
    const courseId: string = 'listWithoutSection';
    const courseName: string = 'List Without Section';
    const students: Student[] =
      [
        {
          courseId,
          email: 'student1OfTypicalCourse@typicalCourse.com',
          name: 'student1OfTypicalCourse',
          joinState: JoinState.JOINED,
          teamName: 'Team 1',
          sectionName: 'None',
        },
        {
          courseId,
          email: 'student2OfTypicalCourse@typicalCourse.com',
          name: 'student2OfTypicalCourse',
          joinState: JoinState.JOINED,
          teamName: 'Team 2',
          sectionName: 'None',
        },
      ];
    expect(service.processStudentsToCsv(courseId, courseName, students)).toMatchSnapshot();
  });

  it('should show course student list with special team name', () => {
    const courseId: string = 'listWithSpecialTeamName';
    const courseName: string = 'List With Special Team Name';
    const students: Student[] =
      [
        {
          courseId,
          email: 'student1OfTypicalCourse@typicalCourse.com',
          name: 'student1OfTypicalCourse',
          joinState: JoinState.JOINED,
          teamName: 'N/A',
          sectionName: 'Section A',
        },
        {
          courseId,
          email: 'student2OfTypicalCourse@typicalCourse.com',
          name: 'student2OfTypicalCourse',
          joinState: JoinState.JOINED,
          teamName: '-Nil-',
          sectionName: 'Section A',
        },
      ];
    expect(service.processStudentsToCsv(courseId, courseName, students)).toMatchSnapshot();
  });

  it('should show course student list with student last name', () => {
    const courseId: string = 'listWithLastName';
    const courseName: string = 'List With Last Name';
    const students: Student[] =
      [
        {
          courseId,
          email: 'student1OfTypicalCourse@typicalCourse.com',
          name: 'student1OfTypicalCourse',
          lastName: ' of The Last Name',
          joinState: JoinState.JOINED,
          teamName: 'Team 1',
          sectionName: 'Section A',
        },
        {
          courseId,
          email: 'student2OfTypicalCourse@typicalCourse.com',
          name: 'student2OfTypicalCourse',
          lastName: "with apostrophe' here",
          joinState: JoinState.JOINED,
          teamName: 'Team 2',
          sectionName: 'Section A',
        },
      ];
    expect(service.processStudentsToCsv(courseId, courseName, students)).toMatchSnapshot();
  });

  it('should show course student list with unregistered student', () => {
    const courseId: string = 'listWithUnregistered';
    const courseName: string = 'List With Unregistered Student';
    const students: Student[] =
      [
        {
          courseId,
          email: 'student1OfTypicalCourse@typicalCourse.com',
          name: 'student1OfTypicalCourse',
          joinState: JoinState.JOINED,
          teamName: 'Team 1',
          sectionName: 'Section A',
        },
        {
          courseId,
          email: 'student2OfTypicalCourse@typicalCourse.com',
          name: 'student2OfTypicalCourse',
          joinState: JoinState.NOT_JOINED,
          teamName: 'Team 2',
          sectionName: 'Section A',
        },
      ];
    expect(service.processStudentsToCsv(courseId, courseName, students)).toMatchSnapshot();
  });
});
