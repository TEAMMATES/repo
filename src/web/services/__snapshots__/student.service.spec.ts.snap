// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StudentService should show course student list with section 1`] = `
"Course ID:,listWithSection
Course Name:,List With Section

Section,Team,Full Name,Last Name,Status,Email
Section A,Team 1,student1OfTypicalCourse,,Joined,student1OfTypicalCourse@typicalCourse.com
Section A,Team 2,student2OfTypicalCourse,,Joined,student2OfTypicalCourse@typicalCourse.com"
`;

exports[`StudentService should show course student list with special team name 1`] = `
"Course ID:,listWithSpecialTeamName
Course Name:,List With Special Team Name

Section,Team,Full Name,Last Name,Status,Email
Section A,-Nil-,student2OfTypicalCourse,,Joined,student2OfTypicalCourse@typicalCourse.com
Section A,N/A,student1OfTypicalCourse,,Joined,student1OfTypicalCourse@typicalCourse.com"
`;

exports[`StudentService should show course student list with student last name 1`] = `
"Course ID:,listWithLastName
Course Name:,List With Last Name

Section,Team,Full Name,Last Name,Status,Email
Section A,Team 1,student1OfTypicalCourse, of The Last Name,Joined,student1OfTypicalCourse@typicalCourse.com
Section A,Team 2,student2OfTypicalCourse,with apostrophe' here,Joined,student2OfTypicalCourse@typicalCourse.com"
`;

exports[`StudentService should show course student list with unregistered student 1`] = `
"Course ID:,listWithUnregistered
Course Name:,List With Unregistered Student

Section,Team,Full Name,Last Name,Status,Email
Section A,Team 1,student1OfTypicalCourse,,Joined,student1OfTypicalCourse@typicalCourse.com
Section A,Team 2,student2OfTypicalCourse,,Yet to Join,student2OfTypicalCourse@typicalCourse.com"
`;

exports[`StudentService should show course student list without section 1`] = `
"Course ID:,listWithoutSection
Course Name:,List Without Section

,Team,Full Name,Last Name,Status,Email
,Team 1,student1OfTypicalCourse,,Joined,student1OfTypicalCourse@typicalCourse.com
,Team 2,student2OfTypicalCourse,,Joined,student2OfTypicalCourse@typicalCourse.com"
`;
